src/basedtyping/__init__.py

```diff
@@ -591,6 +587,116 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
         @deco
         def foo(): ...
     """
-    if not isinstance(fn, FunctionType):  # type: ignore[redundant-expr]
+    if not isinstance(fn, FunctionType):
         raise TypeError(f"{fn} is not a FunctionType")
-    return fn  # type: ignore[unreachable]
+    # https://github.com/KotlinIsland/basedmypy/issues/745
+    return cast("FunctionType[P, T]", fn)
```

Comment:
this isn't a type-guard, it's a type-assertion. the `isinstance` there should have worked as the guard, but it's broken in bmp unfortunately

Reply:
> this isn't a type-guard, it's a type-assertion

I was thinking that you could create a type-guard elsewhere, and use that instead of this type assertion. 
But I'm not sure whether that'll work for the generic bits as well.